### PR TITLE
Move `cabi_realloc` to a `wit-bindgen-rt` crate

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit extracts out a `wit-bindgen-rt` crate which unconditionally provides the `cabi_realloc` function. This'll exist for `cargo component` to add a dependency on and otherwise the `wit-bindgen` crate isn't actually need at runtime at all.

### CODE REVIEW

File: crates/guest-rust/rt/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+/// the linker. The purpose of this function is to force a reference to the
+/// symbol `cabi_realloc` to make its way through to the final linker
+/// command line. That way `wasm-ld` will pick it up, see it needs to be
+/// exported, and then export it.
+///
+/// For more information about this see `./ci/rebuild-libcabi-realloc.sh`.
+pub fn maybe_link_cabi_realloc() {
+    #[cfg(target_family = "wasm")]
+    {
+        #[cfg(feature = "realloc")]
```
Comment (peterhuene) [inline crates/guest-rust/rt/src/lib.rs:16]: I think this feature check needs to be removed, right?

File: crates/guest-rust/rt/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        // but the linker will have at that point seen the `cabi_realloc`
+        // symbol and it should get exported.
+        #[cfg(feature = "realloc")]
+        #[used]
+        static _NAME_DOES_NOT_MATTER: unsafe extern "C" fn(
+            *mut u8,
+            usize,
+            usize,
+            usize,
+        ) -> *mut u8 = cabi_realloc;
```
Comment (peterhuene) [inline crates/guest-rust/rt/src/lib.rs:39]:
```suggestion
```

**Review by peterhuene — APPROVED** [review]
Otherwise, looks good, thanks for the extremely fast turn around!

File: crates/guest-rust/rt/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            ) -> *mut u8;
+        }
+        // Force the `cabi_realloc` symbol to be referenced from here. This
+        // is done with a `#[used]` Rust `static` to ensure that this
+        // reference makes it all the way to the linker before it's
+        // considered for garbage collection. When the linker sees it it'll
+        // remove this `static` here (due to it not actually being needed)
+        // but the linker will have at that point seen the `cabi_realloc`
+        // symbol and it should get exported.
+        #[cfg(feature = "realloc")]
```
Comment (peterhuene) [inline crates/guest-rust/rt/src/lib.rs:32]:
```suggestion
```

File: crates/guest-rust/rt/src/lib.rs (hunk 1 above)
Comment (alexcrichton) [reply crates/guest-rust/rt/src/lib.rs]: Whoops, yes!

**Review by peterhuene — APPROVED** [review]

**Pull request merged**